### PR TITLE
ROC-1521: CCJ defendant date of birth

### PR DESCRIPTION
**Pull request by rattur**
### JIRA link ###

https://tools.hmcts.net/jira/browse/ROC-1531

### Change description ###

Initial components for CCJ defendant date of birth

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/main/features/ccj/views/date-of-birth.njk [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        ],
+        form,
+        isInline = false
+        )
+        }}
+
+        <div class="panel panel-border-narrow js-hidden" id="hasDateOfBirth-true">
+          {{ dateInput('date', form) }}
+        </div>
+
```
Comment (timja-kainos) [inline src/main/features/ccj/views/date-of-birth.njk]: remove extra lines below here

File: src/main/features/ccj/views/date-of-birth.njk [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        <h1 class="heading-large">
+          <span class="heading-secondary">{{ t('County Court Judgment') }}</span>
+          {{ t("Defendant's date of birth") }}
+        </h1>
+
+        <p>{{ t("We use this to ensure legal proceedings are made against the right person. " +
+        "For example, if there are 2 people living at the same address with the same " +
+        "name.") }}</p>
+
+        {{ radioGroup(
```
Comment (timja-kainos) [inline src/main/features/ccj/views/date-of-birth.njk]: is there any validation forcing the user to pick yes or no here?

File: src/test/features/ccj/routes/date-of-birth.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import { app } from '../../../../main/app'
+
+import * as idamServiceMock from '../../../http-mocks/idam'
+import * as claimStoreServiceMock from '../../../http-mocks/claim-store'
+import * as draftStoreServiceMock from '../../../http-mocks/draft-store'
+import { checkAuthorizationGuards } from './checks/authorization-check'
+
+const cookieName: string = config.get<string>('session.cookieName')
+
+const dateOfBirthPage = '/case/b17af4d2-273f-4999-9895-bce382fa24c8/ccj/date-of-birth'
```
Comment (timja-kainos) [inline src/test/features/ccj/routes/date-of-birth.ts]: retrieve from paths and get the external id from the sampleClaimObj

File: src/main/features/ccj/views/date-of-birth.njk (hunk 2 above)
Comment (rattur) [inline src/main/features/ccj/views/date-of-birth.njk]: When page loads, both yes and no are deselected, therefore, asking user to select one or the other

File: src/test/features/ccj/routes/date-of-birth.ts (hunk 3 above)
Comment (rattur) [inline src/test/features/ccj/routes/date-of-birth.ts]: What is the sampleClaimObj?

Comment (rattur) [inline src/test/features/ccj/routes/date-of-birth.ts]: It's ok I found it

File: src/main/features/ccj/views/date-of-birth.njk (hunk 2 above)
Comment (timja-kainos) [inline src/main/features/ccj/views/date-of-birth.njk]: but is it enforced that the user selects one?

Comment (rattur) [inline src/main/features/ccj/views/date-of-birth.njk]: Actually your right, it isn't forcing the user to select one, so I'll look into it

Comment (timja-kainos) [inline src/main/features/ccj/views/date-of-birth.njk]: Have a look at patryk's PR

File: src/main/app/forms/models/dateOfBirth.ts [hunk 4]
```diff
@@ -16,23 +16,27 @@ export class ValidationErrors {
 }
 
 export default class DateOfBirth implements Serializable<DateOfBirth>, CompletableTask {
+  @IsDefined ({ message: 'Choose option: Yes or Not sure' })
```
Comment (geek-soft) [inline src/main/app/forms/models/dateOfBirth.ts]: Worth confirming content with @BenClancy

File: src/main/app/forms/models/dateOfBirth.ts [hunk 5]
```diff
@@ -16,23 +16,27 @@ export class ValidationErrors {
...
+    this.known = known
     this.date = date
   }
 
   static fromObject (value?: any): DateOfBirth {
     if (!value) {
       return value
     }
-    return new DateOfBirth(LocalDate.fromObject(value.date))
+    return new DateOfBirth(Object.getOwnPropertyDescriptor(value, 'known') !== undefined ? value.known === 'true' : undefined, LocalDate.fromObject(value.date))
```
Comment (geek-soft) [inline src/main/app/forms/models/dateOfBirth.ts]: Can we simplify it to be just `value.known === 'true'`. I do not think the longer version does anything for us at the moment due to the way form values are deserialised upon submission.

File: src/main/features/ccj/views/date-of-birth.njk [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        <fieldset class="form-group {% if errorType %} form-group-error {% endif %}">
+
+          <legend>
+              <span class="form-label-bold">
+                {{ t("Do you know the defendant's date of birth?") }}
+              </span>
+          </legend>
+
+          {% if errorType %}<span class="error-message">{{ t(errorType) }}</span>{% endif %}
+          <div class="multiple-choice" data-target="type-known">
```
Comment (geek-soft) [inline src/main/features/ccj/views/date-of-birth.njk]: It should be possible to use `radioOption` here.

File: src/test/features/ccj/routes/date-of-birth.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+  describe('on GET', () => {
+    checkAuthorizationGuards(app, 'get', dateOfBirthPage)
+
+    context('when user authorised', () => {
+      beforeEach(() => {
+        idamServiceMock.resolveRetrieveUserFor(1, 'cmc-private-beta')
+      })
+
+      it('should render page when everything is fine', async () => {
```
Comment (geek-soft) [inline src/test/features/ccj/routes/date-of-birth.ts]: Missing tests for:
- failure in retrieval CCJ draft
- eligible CCD guard

The same comment applies for `POST` endpoint.

File: src/main/app/forms/models/dateOfBirth.ts [hunk 8]
```diff
@@ -16,23 +16,27 @@ export class ValidationErrors {
 }
 
 export default class DateOfBirth implements Serializable<DateOfBirth>, CompletableTask {
+  @IsDefined ({ message: 'Choose option: Yes or Not sure' })
+  known: boolean
 
+  @ValidateIf(o => o.known === true)
```
Comment (geek-soft) [inline src/main/app/forms/models/dateOfBirth.ts]: The new behaviour should be reflected in unit tests. Please update `app/forms/models/dateOfBirth.ts`

**Pull request merged**